components/fxa-client/src/oauth.rs

```diff
@@ -172,25 +172,47 @@ impl FirefoxAccount {
         &self,
         auth_params: AuthorizationParameters,
     ) -> Result<String> {
-        // TODO: Add validation to check for which keys the client is allowed to request,
-        //       And if we have those keys stored locally.
-        //       Currently we pull ALL available keys that match the requested scopes, regardless
-        //       if the client is actually allowed to request them.
         let session_token = self.get_session_token()?;
+
+        // Validate request to ensure that the client is actually allowed to request
+        // the scopes they requested
+        let valid_scopes = self.client.scoped_key_data(
+            &self.state.config,
+            &session_token,
+            &auth_params.client_id,
+            &auth_params.scope,
+        )?;
+
+        if let Some(not_allowed_scope) = auth_params
+            .scope
+            .split_whitespace()
```

Comment:
I wonder if it would be simpler/safer to do this `split_whitespace` step when you're constructing `AuthorizationParameters` struct. That is, have `AuthorizationParameters::scope` be a `Vec<String>` rather than a `String`, so that we don't accidentally forget to call `split_whitespace` if we end up using it in other places in future.

Not a blocker, just a thought.